Add Jetpack Affiliate package

eliorivero opened the pull request:
This PR will create a package for the class that reads and Affiliate code. It's later replaced where it was previously used, to add the affiliate code to the connection URL.

To be merged in https://github.com/Automattic/jetpack/pull/12527

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Move the `Jetpack_Affiliate` class to its own Composer package.
* in files where this class was used, import the namespaced class
* small tweaks in Jetpack_Affiliate class so it's PHPCS compliant

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* checkout the branch
* do `composer require automattic/jetpack-affiliate`
* do `wp option add jetpack_affiliate_code abc123` (or `yarn docker:wp option add jetpack_affiliate_code abc123`)
* go to WP Admin > Dashboard in a _disconnected_ site 
* verify that the URL of the Set up Jetpack button has the `aff` parameter with the right value

#### Proposed changelog entry
* Move Jetpack Affiliate code to a Composer package.

eliorivero committed:
Introduce Affiliate package

tyxla commented on _inc/lib/admin-pages/class.jetpack-react-page.php:
```diff
@@ -1,4 +1,6 @@
 <?php
+use Automattic\Jetpack\Partners\Affiliate as Affiliate;
```
I think ` as Affiliate` isn't necessary, is it? If we omit it, we should end up with the same name.

eliorivero replied:
Fixed. Thanks.

tyxla commented on packages/affiliate/composer.json:
```diff
@@ -0,0 +1,25 @@
+	"description": "Reads an affiliate code for Jetpack. Useful to add it to the connection URL so it's read by wpcom.",
+	"type": "library",
+	"license": "GPL-2.0-or-later",
+	"version": "1.0.0",
```
Could we omit the version? Ideally, once this is published as a proper package, its corresponding GH repo release tags should be the canonical source of truth.

eliorivero replied:
Done.

tyxla commented on packages/affiliate/tests/php/test_Affiliate.php:
```diff
@@ -1,23 +1,20 @@
+use PHPUnit\Framework\TestCase;
 
 // Load required class to get the affiliate code
 require_once JETPACK__PLUGIN_DIR . 'class.jetpack.php';
```
`JETPACK__PLUGIN_DIR` won't be defined here, as these tests run standalone.

eliorivero replied:
Replaced with Constants package.

tyxla commented on packages/affiliate/src/Affiliate.php:
```diff
@@ -68,11 +76,13 @@ public function get_affiliate_code() {
 }
 
-add_action( 'init', array( 'Jetpack_Affiliate', 'init' ) );
+add_action( 'init', array( 'Automattic\Jetpack\Partners\Affiliate', 'init' ) );
```
Ideally, this file should only contain the class definition, and no other definitions or executable code. This action should be hooked somewhere else - ideally, where the library is consumed, so it's more obvious.

eliorivero replied:
Makes sense. Will initialize class in file where this is used.

tyxla commented on packages/affiliate/tests/php/test_Affiliate.php:
```diff
@@ -1,23 +1,20 @@
+use PHPUnit\Framework\TestCase;
 
 // Load required class to get the affiliate code
 require_once JETPACK__PLUGIN_DIR . 'class.jetpack.php';
```
Having this as a requirement for testing indicates that this package isn't independent enough to be a package of its own. Ideally, it should be able to live and run outside of Jetpack, and all of its dependencies should be defined explicitly in its `composer.json`.

tyxla commented on packages/affiliate/src/Affiliate.php:
```diff
@@ -1,25 +1,33 @@
-if ( ! defined( 'ABSPATH' ) ) {
-	// Exit if accessed directly
-	exit;
+use Jetpack;
```
As far as I understand, a package shouldn't do this - packages are expected to be independently distributable packages that don't need Jetpack to work.

This also signifies that there's additional functionality that might make sense to be ported to this package, or to a separate one.

eliorivero replied:
Thanks. Replaced using the Status package.

tyxla commented on packages/affiliate/src/Affiliate.php:
```diff
@@ -1,25 +1,33 @@
-	exit;
+use Jetpack;
+
+if ( ! defined( 'ABSPATH' ) || ! is_admin() ) {
```
These statements shouldn't be necessary, as ideally such files are expected to only contain declarations of classes.

eliorivero replied:
Right.

tyxla commented on packages/affiliate/src/Affiliate.php:
```diff
@@ -1,25 +1,33 @@
+ * - a `jetpack_affiliate_code` option in the WP database
  * - an affiliate code returned by a filter bound to the `jetpack_affiliate_code` filter hook
  *
  * @since 6.9.0
```
Noting that once we move this code to a package, such `@since` references become ambigious, because it's not certain whether this is the Jetpack version or the package version.

eliorivero replied:
All removed.

tyxla commented on packages/affiliate/composer.json:
```diff
@@ -0,0 +1,25 @@
+	"require": {},
+	"require-dev": {
+		"phpunit/phpunit": "^5.7 || ^6.5 || ^7.5",
+		"php-mock/php-mock": "^2.1"
```
It doesn't seem like we're using `phpmock` in this package?

eliorivero replied:
Good catch.

tyxla reviewed (changes requested):
This is a great start 👍 

Unfortunately, I'm getting an `Fatal error: Uncaught Error: Class 'Automattic\Jetpack\Partners\Affiliate' not found in /var/www/html/wp-content/plugins/jetpack/class.jetpack.php on line 4562.` error here. Seems to be present in the integration tests too, as a main reason for their failure.

Package tests also seem to be failing, due to the fact we rely on undefined constants to be defined in our tests (`JETPACK__PLUGIN_DIR`).

eliorivero committed:
Update based on feedback

eliorivero committed:
Update composer lock

eliorivero committed:
Remove unnecessary line loading the old file

eliorivero committed:
Add local repositories to load constants and status packages

eliorivero committed:
Set correct plugin file constant

eliorivero committed:
Mock get_option and apply_filters functions.

Props @zinigor for noticing that the condition checking whether the site was in development mode wasn't really useful. That allowed to remove dependency on Status and clean up the tests.

roccotripaldi reviewed (approved):
Worked well in my tests, and the code looks good.
Thanks for the detailed testing instructions.

tyxla commented on packages/affiliate/README.md:
```diff
@@ -0,0 +1,15 @@
+
+### Usage
+
+Display the default Jetpack logo:
```
Seems like we should change this line to reflect what the example is doing.

tyxla commented on composer.lock:
```diff
@@ -4,15 +4,49 @@
         "This file is @generated automatically"
     ],
-    "content-hash": "c09d9b77fbe174fad981ff2c454e926b",
+    "content-hash": "7e2d82d82702db136be13cf61d05e472",
```
Seems like we'll need to rebase `composer.lock`.

tyxla commented on packages/affiliate/tests/php/test_Affiliate.php:
```diff
@@ -0,0 +1,94 @@
+<?php
+namespace Automattic\Jetpack\Partners;
+
+use Automattic\Jetpack\Constants as Jetpack_Constants;
```
We don't seem to be using this one.

```suggestion
```

tyxla commented on packages/affiliate/src/Affiliate.php:
```diff
@@ -1,58 +1,47 @@
-	 * @since 6.9.0
-	 *
-	 * @return Jetpack_Affiliate | false
+	 * @return Affiliate | false
```
AFAIK it's more common to use the fully qualified class name in the docblocks.

tyxla replied:
Will this ever return `false`?

tyxla commented on packages/affiliate/composer.json:
```diff
@@ -0,0 +1,33 @@
+	"license": "GPL-2.0-or-later",
+	"require": {
+		"automattic/jetpack-constants": "@dev",
+		"automattic/jetpack-status": "@dev"
```
We don't seem to be using any of these 2 packages in this one. Remove and update composer.lock?

tyxla commented on packages/affiliate/src/Affiliate.php:
```diff
@@ -1,58 +1,47 @@
 <?php
```
We'll need to keep the original file for compatibility reasons. You can an example on how to proceed with other classes that have been moved to packages.

Should we also add it to the compat package?

tyxla commented on packages/affiliate/tests/php/test_Affiliate.php:
```diff
@@ -0,0 +1,94 @@
+			]
+		] );
+
+		$this->mock_function_with_args( 'apply_filters', [
```
It appears that we're doing this one for all tests, so I wonder if we could move it to a helper method.

tyxla reviewed (changes requested):
Nice work!

I think we should take care of compatibility with the old class, plus several more smaller things.
